src/builtins.rs

```diff
@@ -0,0 +1,686 @@
+        NixValue::Lambda(_) => Ok(Gc::new(NixValue::Bool(true))),
+        _ => Ok(Gc::new(NixValue::Bool(false))),
+    }
+    "isList" ; IsList => |param: Gc<Tree>| match param.eval()?.borrow() {
+        NixValue::List(_) => Ok(Gc::new(NixValue::Bool(true))),
+        _ => Ok(Gc::new(NixValue::Bool(false))),
+    }
+    "genList" ; GenList => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::GenList1(param)))))
+    }
+    "genList <?>" ; GenList1(_0: Gc<Tree>) => |param: Gc<Tree>, lambda: &Gc<Tree>| {
+        let num = param.eval()?.as_int()?;
+        let mut out = vec![];
+        for i in 0..num {
+            let tree = Gc::new(Tree {
+                value: GcCell::new(None),
+                hash: GcCell::new(None),
+                source: TreeSource::Apply {
+                    lambda: Ok(lambda.clone()),
+                    arg: Ok(Gc::new(Tree::from_concrete(NixValue::Integer(i)))),
+                },
+                range: None,
+                scope: Gc::new(Scope::None)
+            });
+            out.push(tree);
+        }
+        Ok(Gc::new(NixValue::List(out)))
+    }
+    "listToAttrs" ; ListToAttrs => |param: Gc<Tree>| {
+        match param.eval()?.borrow() {
+            NixValue::List(list) => {
+                let mut out_map = HashMap::new();
+                for item in list {
+                    match item.eval()?.borrow() {
+                        NixValue::Map(ref map) => {
+                            let tmp = map.get("name").ok_or(EvalError::Parsing)?;
+                            let key = match tmp.eval()?.borrow() {
+                                NixValue::Str(ref x) => x.clone(),
+                                x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+                            };
+                            let val = map.get("value").ok_or(EvalError::Parsing)?;
+                            out_map.insert(key.to_string(), val.clone());
+                        }
+                        x => return Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+                    }
+                }
+                Ok(Gc::new(NixValue::Map(out_map)))
+            }
+            x => Err(EvalError::Unimplemented(format!("cannot cast {:?}", x))),
+        }
+    }
+    "split" ; Split => |param: Gc<Tree>| {
+        Ok(Gc::new(NixValue::Lambda(NixLambda::Builtin(NixBuiltin::Split1(param.eval()?.as_str()?)))))
+    }
+    "split <?>" ; Split1(_0: String) => |param: Gc<Tree>, regex_str: &String| {
+        let regex_str = regex_str
+            .replace("\\|", "%%PIPE%%")
+            .replace("\\(", "%%L_PAREN%%")
+            .replace("\\)", "%%R_PAREN%%")
+            .replace("|", "\\|")
+            .replace("(", "\\(")
+            .replace(")", "\\)")
+            .replace("%%PIPE%%", "|")
+            .replace("%%L_PAREN%%", "(")
+            .replace("%%R_PAREN%%", ")");
```

Comment:
Would you mind elaborating here what's this about? :)

Reply:
Ah this is a very hacky way to handle this. I don't want this hack in rnix-lsp, so I'll need to replace it before putting it in any PRs I want merged